[RFC] Allow configuring storage options for keyspaces

psarna opened the pull request:
The STORAGE option is designed to hold a map of options
used for customizing storage for given keyspace.

Example:
```cql
cassandra@cqlsh> CREATE KEYSPACE ks WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 } AND STORAGE = { 'type' : 'S3', 'bucket' : '42', 'key_id' : '43', 'endpoint' : 'localhost' } ;
cassandra@cqlsh> select * from system_schema.keyspaces ;

 keyspace_name                 | durable_writes | replication                                                                         | storage
-------------------------------+----------------+-------------------------------------------------------------------------------------+-------------------------------------------------------------------------
                   system_auth |           True | {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '1'} |                                                      {'type': 'NATIVE'}
                 system_schema |           True |                             {'class': 'org.apache.cassandra.locator.LocalStrategy'} |                                                      {'type': 'NATIVE'}
            system_distributed |           True | {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '3'} |                                                      {'type': 'NATIVE'}
                        system |           True |                             {'class': 'org.apache.cassandra.locator.LocalStrategy'} |                                                      {'type': 'NATIVE'}
                            ks |           True | {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '1'} | {'bucket': '42', 'endpoint': 'localhost', 'key_id': '43', 'type': 'S3'}
                 system_traces |           True | {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '2'} |                                                      {'type': 'NATIVE'}
 system_distributed_everywhere |           True |                        {'class': 'org.apache.cassandra.locator.EverywhereStrategy'} |                                                      {'type': 'NATIVE'}

```

RFC because:
 1. the storage options are not used anywhere
 2. no tests are added yet (a cql-pytest case will be added)
 3. the commit absolutely disregards the fact that `system_schema.keyspaces` just got a new column, which isn't backwrad compatible

I'd especially appreciate a suggestion for solving (3.) - should we extend `system_schema.keyspaces`, and follow with introducing a schema feature + some more complicated backward compatibility tests? One alternative is a separate table (`scylla_keyspaces`?), but I'm not sure if there's any real gain, maybe except keeping the original schema tables compatible with Cassandra.

psarna committed:
database,cql3: add STORAGE option to keyspaces

The STORAGE option is designed to hold a map of options
used for customizing storage for given keyspace.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1762,7 +1768,16 @@ lw_shared_ptr<keyspace_metadata> create_keyspace_from_schema_partition(const sch
+    if (storage) {
+        for (auto& p : *storage) {
+            storage_options.emplace(value_cast<sstring>(p.first), value_cast<sstring>(p.second));
+        }
```
Let's define the actual storage types:
 - provider: s3/filesystem/etc.
 - credentials: map? what?
 - bucket: bucket to store the data in

psarna replied:
right, and if we migrate from keeping everything in a map, it makes even more sense to introduce a new schema table - keyspace_storage, which all of these as fields

avikivity commented:
We can keep using system_schema.keyspaces, it's fine.

psarna commented:
> We still need a way to store the bucket, access point, and credentials.

Yeah, I'll add it later today when I'm back coding. Meanwhile, if anybody has an idea how to represent a particular attribute (e.g. what's a bucket, a string?), feel free to post comments here, it will make my job easier later.

psarna committed:
cql-pytest: add test for keyspace storage options

A minimal test case verifies that storage options
set up during keyspace creation are persisted.
